### PR TITLE
PDF stamping of approved invoices 

### LINKED ISSUE #3539: Invoice has name of approver and date stamp
### Description

Manipulate PDF invoice submitted by vendor with the approver information and date stamp

### PR DESCRIPTION
**Pull request by theskumar**
- [x] Add approved by staff to the approved invoice pdfs
- [x] Add inovice approver and date

Screenshot 2024-05-20 at 6  33 05@2x

Closes: https://github.com/HyphaApp/hypha/issues/3539

### CODE REVIEW

**Comment by theskumar** [comment]
@sandeepsajan0 @frjo Are you good with denormalizing and add the `approved_at` and `approved_by` to the invoice model? 

The code to update them will go in here: 

https://github.com/HyphaApp/hypha/blob/1b401334914f013b7d30e9bf4c2838361ee7c8f7/hypha/apply/projects/views/payment.py#L81-L87

**Comment by theskumar** [comment]
Added the logic to store the approver and date at https://github.com/HyphaApp/hypha/pull/3698/commits/f426d46f5be088b3228dca94c155789ceb28128c

Let me know if this something that need to be improved upon? e.g. displaying both finance1 and finance2 info etc.

File: hypha/apply/projects/models/payment.py
Comment (sandeepsajan0) [inline hypha/apply/projects/models/payment.py]: Can we rename it to `final_approval_...` or something that doesn't confuse it with internal approvals like staff approval or finance approval(in case of extended workflow)? Because we have to stamp the pdf only with final approval(finance approval or finance2 approval).

Or we may directly look out for the final approval activity as the [activity](https://github.com/HyphaApp/hypha/blob/main/hypha/apply/activity/models.py#L98) already has the date and user.

File: hypha/apply/projects/views/payment.py
Comment (sandeepsajan0) [inline hypha/apply/projects/views/payment.py]: It is not the right place to update approved_by or approved_at because here we are just notifying the finance members to approve the invoice.

It requires adding another condition to check if the status is `Approved by Finance`(without extended).

**Comment by theskumar** [comment]
@sandeepsajan0 thanks for feedback. I would like wait for the feedback on the content sample itself and then tackle both at once. 

- https://github.com/HyphaApp/hypha/issues/3539#issuecomment-1860480909

**Comment by theskumar** [comment]
Updated the code to make use of activity feed. It will list all the approvals with timestamps.

File: hypha/apply/projects/views/payment.py
Comment (sandeepsajan0) [inline hypha/apply/projects/views/payment.py:436]: It will pick all the Approved by ... statuses. If an invoice has been back and forth between finance and staff then it might show multiple `Approved by Staff` but I think only latest one should be fine. We might need distinct activities.  Also, we don't want to add this for invoices with status `Approved by finance`?  And we might need to hide it from the applicant.
image

File: hypha/apply/projects/views/payment.py
Comment (sandeepsajan0) [inline hypha/apply/projects/views/payment.py:126]: I think it was getting saved in `form_valid()` in line100?

**Review by wes-otf — APPROVED** [review]

**Pull request merged**
